src/NuGet.Core/NuGet.Common/Migrations/MigrationRunner.cs

```diff
@@ -32,31 +32,42 @@ public static void Run()
                 // so use a global mutex and then check if someone else already did the work.
                 using (var mutex = new Mutex(false, "NuGet-Migrations"))
                 {
-                    bool signal = mutex.WaitOne(TimeSpan.FromMinutes(1), false);
-                    if (signal && !File.Exists(expectedMigrationFilename))
+                    bool captured;
+                    try
```

Comment:
I'm not sure I understand why the try catch is not wrapping the complete method below. 

Is there a particular reason for that? 
imo, it'd be cleaner that way, but maybe you had a reason.

Reply:
I just pushed a commit introducing this change to address https://github.com/NuGet/NuGet.Client/pull/4859/files/1f4771d337b4e4eba1d7360715fa883a93234f37#r1000018933 comment. Happy to get your feedback and adjust functionality.

Reply:
I think my suggestion is 

```cs
bool captured;
                    try
                    {
                        if (!File.Exists(expectedMigrationFilename))
                        {
                            // Only run migrations that have not already been run
                            int highestMigrationRun = GetHighestMigrationRun(migrationsDirectory);
                            for (int i = highestMigrationRun + 1; i < Migrations.Count; i++)
                            {
                                try
                                {
                                    Migrations[i]();
                                    // Create file for every migration run, so that if an older version of NuGet is run, it doesn't try to run
                                    // migrations again.
                                    string migrationFile = Path.Combine(migrationsDirectory, (i + 1).ToString(CultureInfo.InvariantCulture));
                                    File.WriteAllText(migrationFile, string.Empty);
                                }
                                catch { }
                            }
                        }
                        mutex.ReleaseMutex();
                    }
                    catch (AbandonedMutexException ex)
                    {
                        ex.Mutex?.ReleaseMutex();
                    }

```

Reply:
@zivkan raised a good question in https://github.com/NuGet/NuGet.Client/pull/4859#discussion_r1000018933 comment, saying that `My understanding of the docs is that AbandonedMutexException means that the current process successfully obtained the mutex, just that the other process failed to release it cleanly. If the other process (or thread?) was killed, then migrations might not have completed, so the current thread should probably ensure that all migrations completed successfully. `

 `AbandonedMutexException` is only raised by `Mutex.WaitOne` method. If that method invocation is not in the try block, then we may not fix this actual issue. Hence, I modified the location of catch block in the new commit so that migrations can run one more time incase if the previous thread didn't run successfully.

Reply:
It might be more clear to readers if you extract the try-catch into a method named `WaitForMutex`. Though a comment on the `catch AbandonedMutexException` to explain that it obtained the mutex, despite the exception, would be very helpful too since it's really unintuitive to anyone who isn't familiar with this already.